cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,750 @@
+    uint64_t const full_length_flipped = swap_endian(message_length_in_bits);
+    memcpy(state.buffer + Hasher::message_chunk_size - message_length_supported_size,
+           reinterpret_cast<uint8_t const*>(&full_length_flipped),
+           message_length_supported_size);
+    this->underlying().hash_step(state);
+
+    // Each byte in the word generates two bytes in the hexadecimal string digest.
+    // SHA-224 and SHA-384 digests are truncated because their digest does not
+    // include all of the hash values.
+    auto constexpr num_words_to_copy =
+      Hasher::digest_size / (2 * sizeof(typename Hasher::sha_word_type));
+    for (int i = 0; i < num_words_to_copy; i++) {
+      // Convert word representation from big-endian to little-endian.
+      typename Hasher::sha_word_type flipped = swap_endian(state.hash_value[i]);
+      if constexpr (std::is_same_v<typename Hasher::sha_word_type, uint32_t>) {
+        uint32ToLowercaseHexString(flipped, result_location + (8 * i));
+      } else if constexpr (std::is_same_v<typename Hasher::sha_word_type, uint64_t>) {
+        uint32_t low_bits = static_cast<uint32_t>(flipped);
+        uint32ToLowercaseHexString(low_bits, result_location + (16 * i));
+        uint32_t high_bits = static_cast<uint32_t>(flipped >> 32);
+        uint32ToLowercaseHexString(high_bits, result_location + (16 * i) + 8);
+      } else {
+        cudf_assert(false && "Unsupported SHA word type.");
+      }
+    }
+  };
+};
+
+template <typename Hasher>
+struct HasherDispatcher {
+  Hasher* hasher;
+  column_device_view col;
+
+  template <typename T,
+            typename std::enable_if_t<(!is_fixed_width<T>() || is_chrono<T>()) &&
+                                      !std::is_same_v<T, string_view>>* = nullptr>
+  void CUDA_DEVICE_CALLABLE operator()(size_type)
+  {
+    cudf_assert(false && "Unsupported type for hash function.");
+  }
+
+  template <typename T,
+            typename std::enable_if_t<is_fixed_width<T>() && !is_chrono<T>()>* = nullptr>
+  void CUDA_DEVICE_CALLABLE operator()(size_type row_index)
+  {
+    T const& key = col.element<T>(row_index);
+    if constexpr (is_floating_point<T>()) {
+      if (isnan(key)) {
+        T nan = std::numeric_limits<T>::quiet_NaN();
+        hasher->process_fixed_width(nan);
+      } else if (key == T{0.0}) {
+        hasher->process_fixed_width(T{0.0});
+      } else {
+        hasher->process_fixed_width(key);
+      }
+    } else {
+      hasher->process_fixed_width(key);
+    }
+  }
+
+  template <typename T, typename std::enable_if_t<std::is_same_v<T, string_view>>* = nullptr>
+  void CUDA_DEVICE_CALLABLE operator()(size_type row_index)
+  {
+    string_view key     = col.element<string_view>(row_index);
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(key.data());
+    uint32_t const len  = static_cast<uint32_t>(key.size_bytes());
+    hasher->process(data, len);
+  }
+};
+
+struct sha1_hash_state {
```

Comment:
I think it would be cleaner if these were nested classes of their respective hash derived types.

e.g.,

```
struct SHA1Hash : HashBase<SHA1Hash> {
  struct hash_state{
     uint64_t message_length = 0;
     uint32_t buffer_length  = 0;
     uint32_t hash_value[5]  = {0x67452301, 0xefcdab89, 0x98badcfe, 0x10325476, 0xc3d2e1f0};
     uint8_t buffer[64];
   };
   ... 
};